src/jit/decomposelongs.cpp

```diff
@@ -690,6 +690,10 @@ GenTree* DecomposeLongs::DecomposeCnsLng(LIR::Use& use)
     GenTree* tree  = use.Def();
     INT32    hiVal = tree->AsLngCon()->HiVal();
 
+#ifdef FEATURE_CROSSBITNESS
```

Comment:
I don't understand this. Can you please explain why this is required? SetLngValue is already target-specific.

Reply:
We discussed this change and did not find a good solution.
The problem here is that `ChangeOperConst` in this case casts `uin64`  to `size_t` that on x86 platform truncates the high part, because `sizeof(size_t) == 4`. However, when host is 64 bites it does not happen and the low part keeps its big long value that produce incorrect results.

My opinion is that we should make this clearing inside  `ChangeOperConst` for all platforms without `FEATURE_CROSSBITNESS` checks, but it needs additional work and checks.

Reply:
Ok, @sandreenko explained this to me, and it's related to node bashing.

I would suggest changing `GenTree::SetOper`, pseudo-code:
```
#if defined(_HOST_64BIT_) && !defined(_TARGET_64BIT_)
if ((old oper == GT_CNS_LNG) && (new oper == GT_CNS_INT)) {
    // When casting from LONG to INT, we need to force cast of the value,
    // if the host architecture represents INT and LONG with the same data size.    
    gtIntCon.gtIconVal = (ssize_t)(int)gtIntCon.gtIconVal;
}
#endif
```

Reply:
@BruceForstall I did similarly as you suggested and also used you comment